warehouse/macaroons/caveats/__init__.py

```diff
@@ -100,6 +101,38 @@ def verify(self, request: Request, context: Any, permission: str) -> Result:
         return Success()
 
 
+@as_caveat(tag=4)
+@dataclass(frozen=True)
+class OIDCProvider(Caveat):
+    oidc_provider_id: StrictStr
+
+    def verify(self, request: Request, context: Any, permission: str) -> Result:
+        # This caveat can only be used when the surrounding request is identified
+        # with OpenID, rather than a user or some other identity.
+        if not isinstance(request.identity, oidc_models.OIDCProvider):
+            return Failure(
+                "OIDC scoped token used outside of an OIDC identified request"
+            )
+
+        if str(request.identity.id) != self.oidc_provider_id:
+            return Failure(
+                "current OIDC provider does not match provider restriction in token"
+            )
+
+        # OpenID-scoped tokens are only valid against projects.
+        if not isinstance(context, Project):
+            return Failure("OIDC scoped token used outside of a project context")
+
+        # Specifically, they are only valid against projects that are registered
+        # to the current identifying OpenID provider.
+        if context not in request.identity.projects:
+            return Failure(
+                f"OIDC scoped token is not valid for project '{project.name}'"
+            )
+
+        return Success()
```

Comment:
cc @dstufft I'm pretty sure I got this right, but since you designed the new API here I figured I'd flag it specifically for your eyes 🙂